shared/application_mode_managers/progression_mode_manager.ts

```diff
@@ -0,0 +1,304 @@
+import {Subscription} from 'rxjs';
+
+import {jimmySet1, jimmySet2, michaelSet1} from '../constants/progression_constants';
+
+import MidiService, {MidiSubjectMessage} from '../io/midi/midi_service';
+import WledService from '../io/wled/wled_service';
+import {Config} from '../types/config_types/config_types';
+
+import {log} from '../utils';
+
+import type App from '../app';
+import {MidiInstrumentName} from '../constants/midi_instrument_constants';
+import {ProgressionState} from '../state/progression_state';
+import {equalControlButton, equalKeyboard} from '../io/midi/midi_utls';
+
+import {ApplicationModeManager} from './application_mode_manager';
+
+type MidiEventHandler = ((msg: MidiSubjectMessage) => void) | undefined;
+
+const songs: number[][][][] = [
+    // set1,
+    // set2,
+    jimmySet1,
+    jimmySet2,
+    michaelSet1,
+];
+
+export default class ProgressionModeManager implements ApplicationModeManager<ProgressionState> {
+    private actions: MidiEventHandler[];
+    private midiEventHandlers: {[eventName: string]: MidiEventHandler};
+    private progressionState: ProgressionState = {
+        currentChord: 0,
+        currentProgression: 0,
+        currentSong: 0,
+        shouldDrumsChangeColor: true,
+        shouldDrumsChangeProgression: false,
+    };
+
+    private midiServiceSubject: Subscription;
+    // private qwertyServiceSubject: Subscription;
+
+    // private dynamicMapping: BluetoothRemoteDynamicMapping;
+
+    constructor(
+        private midiService: MidiService,
+        private wledService: WledService,
+        // private qwertyService: QwertyService,
+        private config: Config,
+        private app: App,
+    ) {
+        this.actions = [
+            this.app.actions.toggleDrumsMusicAction,
+            this.app.actions.noteOffAll,
+            this.app.actions.setRandomColor,
+            this.app.actions.nextPreset,
+            this.nextSong,
+            this.playChordAndChangeColor,
+            this.playChordAndChangePreset,
+            this.playChordAndChangePreset,
+            this.nextProgressionAndPreset,
+            this.app.actions.setRandomEffect,
+        ];
```

Comment:
Let's define data-driven buttons here. Should be easy to just attach callbacks to the button definitions here

Need to define how to convey the "Main Trigger" in the UI. Maybe there is a "Main Trigger" thing outside of the 2x4 grid of buttons. It should also tell the keyboardist to use the sustain pedal to trigger chords. Maybe also display information on how your keyboard is being used to control another keyboard if that's happening